### PR TITLE
Support new ELF metadata + Add APDU dump with new Ledgerctl option

### PR DESCRIPTION
**Pull request by agrojean-ledger**

### CODE REVIEW

**Review by xchapron-ledger — COMMENTED** [review]
Is is retro compatible with app where the api_level is an int and not a string?
I guess we should try it to avoid issue during deployments until all apps uses the updated SDK.

**Comment by agrojean-ledger** [comment]
> Is is retro compatible with app where the api_level is an int and not a string? I guess we should try it to avoid issue during deployments until all apps uses the updated SDK.

I'll make it retro compatible. Thanks for noticing this potential issue.

**Comment by agrojean-ledger** [comment]
@xchapron-ledger I made sure to be retro compatible :

Tested with :
* both `.ledger.api_level` and `ledger.api_level` (with or without prefix `.`).
* both u8 or string storage for api_level.

File: src/utils.rs [hunk 1]
```diff
@@ -32,10 +50,19 @@ pub fn retrieve_infos(
...
+            if matches!(name, "ledger.api_level" | ".ledger.api_level") {
+                if contains_number_representation(&buffer, &(section.sh_offset as usize)) {
+                    infos.api_level = get_string_from_offset(&buffer, &(section.sh_offset as usize));
+                }
+                else {
+                    // Assume the api_level is stored as a u8
+                    infos.api_level = buffer[section.sh_offset as usize].to_string();
+                }
+                break;
+            }
```
Comment (xchapron-ledger) [inline src/utils.rs:65]: Maybe you could simplify by considering/
- .ledger.api_level = int
- ledger.apli_level = string
?

Comment (agrojean-ledger) [reply src/utils.rs:65]: I find it best not to assume anything about the file we parse ^^

Comment (agrojean-ledger) [reply src/utils.rs]: @xchapron-ledger I have simplified it as you suggested.

File: src/utils.rs [hunk 2]
```diff
@@ -19,13 +30,26 @@ pub fn retrieve_infos(
     let mut infos = LedgerAppInfos::default();
 
     // All infos coming from the SDK are expected to be regrouped
-    // in various `.ledger.<field_name>` section of the binary.
-    // For now we only need the API_LEVEL
+    // in various `.ledger.<field_name>` (rust SDK <= 0.2.1) or
+    // `ledger.<field_name> (rust SDK > 0.2.1) section of the binary.
```
Comment (xchapron-ledger) [inline src/utils.rs:34]: version should be updated.

**Review by xchapron-ledger — APPROVED** [review]

File: src/main.rs [hunk 3]
```diff
@@ -292,11 +293,34 @@ fn build_app(
...
+            } else {
+                remaining_args
+                    .get(index + 1)
+                    .map(|path_str| path_str.to_string())
+            }
+        })
+        .map(|path_str| PathBuf::from(path_str));
+    let exe_filename = exe_path.file_name().unwrap().to_str();
+    let exe_parent = exe_path.parent().unwrap().to_path_buf();
+    let apdu_file_path = output_dir.unwrap_or(exe_parent).join(exe_filename.unwrap()).with_extension("apdu");
```
[the remark this answers is not in the document]
Comment (agrojean-ledger) [reply src/main.rs:321]: Yes for now the name of the .apdu, .elf... is based on the package name.

Until we do the work in cargo ledger, in the VS Code extension I will rename the files so they are easily found like C apps files.

**Review by xchapron-ledger — APPROVED** [review]

**Pull request merged**